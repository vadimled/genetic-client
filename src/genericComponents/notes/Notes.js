import React, { Component, Fragment, createRef } from "react";
import { Tooltip } from "antd";
import style from "./Notes.module.scss";
import { LIMITS, TEXTS } from "Utils/constants";
import { ReactComponent as EditIcon } from "Assets/edit.svg";
import PropTypes from "prop-types";
import EditNotes from "./components/editNotes";

class Notes extends Component {
  constructor(props) {
    super(props);

    this.initialState = {
      isEdit: false,
      editNotes: props.value,
      limit: {
        value: 0
      }
    };

    this.state = {
      ...this.initialState
    };

    this.notes = createRef();

    document.addEventListener("mousedown", this.cleanupOutside, false);
  }

  componentWillUnmount() {
    document.removeEventListener("mousedown", this.cleanupOutside, false);
  }

  validateNotesLength = number => {
    if (number <= LIMITS.maxNotesChar) {
      return {
        validateStatus: "success",
        errorMsg: null
      };
    }
    return {
      validateStatus: "error",
      errorMsg: `Maximum ${LIMITS.maxNotesChar} characters`
    };
  };

  cleanupOutside = e => {
    if (!this.notes.current.contains(e.target)) {
      this.cleanupState();
    }
  };

  handleDone = () => {
    // <<<<<<< HEAD:src/genericComponents/notes/Notes.js
    const { setNotes, updateActivityLog, value: prevValue, tableRow } = this.props;
    setNotes(this.state.editNotes);
    updateActivityLog({
      prevValue,
      item: tableRow,
      changedField: "notes"
    });

    this.cleanupState();
  };

  cleanupState = () => {
    this.setState({ ...this.initialState });
  };

  handelEditClick = () => {
    const { value } = this.props;
    this.setState({ editNotes: value, isEdit: true });
  };

  handleEditNotesOnChange = e => {
    const { value } = e.target;
    let amount = value.length;
    this.setState({
      editNotes: value,
      limit: {
        ...this.validateNotesLength(amount),
        value: amount
      }
    });
  };

  setPopup = () => {
    return (
      <div className="edit-text-box">
        <EditNotes
          doneHandler={this.handleDone}
          cancelHandler={this.cleanupState}
          handleOnChange={this.handleEditNotesOnChange}
          notesValue={this.state.editNotes}
          validateStatus={this.state.limit}
        />
      </div>
    );
  };

  render() {
    const { value, placeholder } = this.props;

    return (
      <div ref={this.notes} className={style["notes-wrapper"]}>
        {!value ? (
          <div className="notes-content-empty" onClick={this.handelEditClick}>
            {placeholder || TEXTS.addNote}
          </div>
        ) : (
          <Fragment>
            <Tooltip placement="topLeft" title={value}>
              <div className="notes-content">{value}</div>
            </Tooltip>

            <div
              data-testid="edit-icon"
              className="notes-icon"
              onClick={this.handelEditClick}>
              <EditIcon />
            </div>
          </Fragment>
        )}
        {this.state.isEdit && this.setPopup()}
      </div>
    );
  }
}

Notes.propTypes = {
  setNotes: PropTypes.func.isRequired,
  updateActivityLog: PropTypes.func,
  value: PropTypes.string,
  tableRow: PropTypes.object,
  placeholder: PropTypes.string
};

Notes.defaultProps = {
  value: "",
  tableRow: {},
  updateActivityLog: () => false
};

export default React.memo(Notes);
