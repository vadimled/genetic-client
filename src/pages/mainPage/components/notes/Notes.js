import React, { Component, Fragment } from "react";
import {connect} from "react-redux";
import { Tooltip } from "antd";
import style from "./Notes.module.scss";
import { LIMITS, TEXTS } from "Utils/constants";
import { ReactComponent as EditIcon } from "Assets/edit.svg";
import PropTypes from "prop-types";
import EditNotes from "Pages/mainPage/components/notes/components/editNotes";
import { setNotes } from "Store/actions/tableActions";

class Notes extends Component {
  constructor(props) {
    super(props);

    this.initialState = {
      isEdit: false,
      editNotes: "",
      limit: {
        value: 0
      }
    };

    this.state = {
      ...this.initialState
    };
  }

  validateNotesLength = number => {
    if (number <= LIMITS.maxNotesChar) {
      return {
        validateStatus: "success",
        errorMsg: null
      };
    }
    return {
      validateStatus: "error",
      errorMsg: `Maximum ${LIMITS.maxNotesChar} characters`
    };
  };

  handleDone = () => {
    this.props.setNotes(this.state.editNotes);
    this.setState({ ...this.initialState });
  };

  handleCancel = () => {
    this.setState({ ...this.initialState });
  };

  handelEditClick = e => {
    e.stopPropagation();
    this.setState({ isEdit: true });
  };

  handleEditNotesOnChange = e => {
    const { value } = e.target;
    let amount = value.length;
    this.setState({
      editNotes: value,
      limit: {
        ...this.validateNotesLength(amount),
        value: amount
      }
    });
  };

  render() {
    const { valueNotes } = this.props;
    return (
      <div className={style["notes-wrapper"]}>
        {!valueNotes ? (
          <div className="notes-content-empty" onClick={this.handelEditClick}>
            {TEXTS.addNote}
          </div>
        ) : (
          <Fragment>
            <Tooltip placement="bottom" title={valueNotes}>
              <div className="notes-content">{valueNotes}</div>
            </Tooltip>
            <div className="notes-icon" onClick={this.handelEditClick}>
              <EditIcon />
            </div>
          </Fragment>
        )}
        {this.state.isEdit && (
          <div className="edit-text-box">
            <EditNotes
              doneHandler={this.handleDone}
              cancelHandler={this.handleCancel}
              handleOnChange={this.handleEditNotesOnChange}
              notesValue={this.state.editNotes}
              validateStatus={this.state.limit}
            />
          </div>
        )}
      </div>
    );
  }
}

Notes.propTypes = {
  valueNotes: PropTypes.string
};

function mapDispatchToProps(dispatch) {
  return {
    setNotes: (data) => dispatch(setNotes(data)),
  };
}

export default connect(null, mapDispatchToProps)(Notes);
