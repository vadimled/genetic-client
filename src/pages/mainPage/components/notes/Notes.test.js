import React from "react";
import { fireEvent } from "@testing-library/react";
import "jest-dom/extend-expect";
import { renderWithRedux } from "Utils/test_helpers";
import Notes from "Pages/mainPage/components/notes/Notes";

describe("Notes", () => {
  let getByTestId, icon, store;

  beforeEach(() => {
    const queries = renderWithRedux(
      <Notes id={0} updateActivityLog={() => {}} />
    );
    getByTestId = queries.getByTestId;
    store = queries.store;
    icon = getByTestId("edit-icon");
  });

  it("if EditIcon clicked", () => {
    fireEvent.click(icon);
    const editTextBox = getByTestId("edit-text-box");
    expect(editTextBox).toBeInTheDocument();

    const done = getByTestId("footer-button-done");
    fireEvent.click(done);
    expect(editTextBox).not.toBeInTheDocument();

    fireEvent.click(icon);
    fireEvent.click(getByTestId("footer-button-cancel"));
    expect(editTextBox).not.toBeInTheDocument();
  });

  it("if edit-text-box opened", () => {
    fireEvent.click(icon);

    const storeNotes = store.getState().table?.data?.[0].notes,
      textEditBox = getByTestId("edit-text-box-textarea").value;

    expect(storeNotes).toEqual(textEditBox);
  });

  it("if TextArea has more then 150 chars", () => {
    fireEvent.click(icon);

    let done = getByTestId("footer-button-done"),
      textEditBox = getByTestId("edit-text-box-textarea"),
      value = "";

    for (let i = 0; i < 151; i++) {
      value += "a";
    }

    fireEvent.change(textEditBox, { target: { value } });
    expect(done).toBeDisabled();
    expect(document.querySelector(".ant-form-explain")).toBeInTheDocument();
  });

  it("if TextArea has less then 150 chars", () => {
    fireEvent.click(icon);

    let done = getByTestId("footer-button-done"),
      textEditBox = getByTestId("edit-text-box-textarea"),
      value = "";

    for (let i = 0; i < 150; i++) {
      value += "a";
    }

    fireEvent.change(textEditBox, { target: { value } });
    expect(done).not.toBeDisabled();
    expect(document.querySelector(".ant-form-explain")).not.toBeInTheDocument();
  });
});
