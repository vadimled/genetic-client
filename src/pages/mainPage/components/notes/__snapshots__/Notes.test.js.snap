// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notes create snapshot 1`] = `
<DocumentFragment>
  <div
    class="notes-wrapper"
  >
    <div
      class="notes-content"
    >
      Eveniet distinctio sit itaque sunt aut qui.
    </div>
    <div
      class="notes-icon"
      data-testid="edit-icon"
    >
      <svg>
        edit.svg
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
